### PR TITLE
8310844: [AArch64] C1 compilation fails because monitor offset in OSR buffer is too large for immediate

### PR DESCRIPTION
**Pull request by TobiHartmann**
[JDK-8287349](https://bugs.openjdk.org/browse/JDK-8287349) changed the code in `LIR_Assembler::osr_entry()` to use a single `ldp` instruction instead of two `ldr` instructions to load the monitor lock and object from the OSR state. This is not correct because the `ldp` instruction only supports a [7-bit signed immediate value](https://developer.arm.com/documentation/den0024/a/The-A64-instruction-set/Memory-access-instructions/Accessing-multiple-memory-locations). If the offset is larger, for example due to a large number of locals as in `TestLargeMonitorOffset::test`, we hit the `Field too big for insn` guarantee.

I suggest to revert [JDK-8287349](https://bugs.openjdk.org/browse/JDK-8287349). 

I also found two unrelated bugs when working on the reproducer: [JDK-8322992](https://bugs.openjdk.org/browse/JDK-8322992) (javac) and [JDK-8322996](https://bugs.openjdk.org/browse/JDK-8322996) (C2).

Thanks,
Tobias

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310844](https://bugs.openjdk.org/browse/JDK-8310844): [AArch64] C1 compilation fails because monitor offset in OSR buffer is too large for immediate (**Bug** - P3)

### Reviewers
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17266/head:pull/17266` \
`$ git checkout pull/17266`

Update a local copy of the PR: \
`$ git checkout pull/17266` \
`$ git pull https://git.openjdk.org/jdk.git pull/17266/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17266`

View PR using the GUI difftool: \
`$ git pr show -t 17266`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17266.diff">https://git.openjdk.org/jdk/pull/17266.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17266#issuecomment-1877039504)

### CODE REVIEW

File: src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp [hunk 1]
```diff
@@ -282,9 +282,10 @@ void LIR_Assembler::osr_entry() {
         __ bind(L);
       }
 #endif
-      __ ldp(r19, r20, Address(OSR_buf, slot_offset));
+      __ ldr(r19, Address(OSR_buf, slot_offset + 0));
       __ str(r19, frame_map()->address_for_monitor_lock(i));
-      __ str(r20, frame_map()->address_for_monitor_object(i));
+      __ ldr(r19, Address(OSR_buf, slot_offset + 1*BytesPerWord));
+      __ str(r19, frame_map()->address_for_monitor_object(i));
     }
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:289]: The macro assembler automagically fuses `ldr` pairs. It'd be better to fix this with:
```

--- a/src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp
+++ b/src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp
@@ -282,7 +282,8 @@ void LIR_Assembler::osr_entry() {
         __ bind(L);
       }
 #endif
-      __ ldp(r19, r20, Address(OSR_buf, slot_offset));
+      __ ldr(r19, Address(OSR_buf, slot_offset));
+      __ ldr(r20, Address(OSR_buf, slot_offset + BytesPerWord));
       __ str(r19, frame_map()->address_for_monitor_lock(i));
       __ str(r20, frame_map()->address_for_monitor_object(i));
     }

```

Comment (TobiHartmann) [reply src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:289]: Thanks for the review. I adjusted the fix accordingly.

**Review by theRealAph — APPROVED** [review]

**Comment by TobiHartmann** [comment]
Thanks for the review, Andrew.

**Comment by theRealAph** [comment]
I looked through the history and I see this bug is my fault, and your fix will have to be back ported to all releases. Argh!
Thanks for fixing it.

**Comment by TobiHartmann** [comment]
But, as I mentioned in the description, it's a regression from [JDK-8287349](https://bugs.openjdk.org/browse/JDK-8287349), right?

**Comment by theRealAph** [comment]
> But, as I mentioned in the description, it's a regression from [JDK-8287349](https://bugs.openjdk.org/browse/JDK-8287349), right?

Yeah, that's true. A "trivial performance fix," as was said at the time. Memo to myself: there are no trivial performance fixes.

File: src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp (hunk 1 above)
Comment (adinn) [reply src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:289]: I'm not sure why the recommended adjustment is needed. The macro assembler does fuse pairs of adjacent ldr instructions into an ldp but only when the sizes match and the offsets fit into the requisite number of bits.

So, if the two ldr instrctions ar egenerated next to each other the macroasembler should only convert to ldp *where appropriate*. Am I missing something here?

Comment (adinn) [reply src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:289]: Doh, sorry - I misread Andrew's proposed code! Ignore the noise.

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:289]: Yes, the problem @TobiHartmann is fixing is that we currently use `ldp`, but in very rare cases`ldp` can't reach, so the fix we need is to change one `ldp` to two `ldr`s. In almost all cases, macroassembler will merge the `ldr`s.

**Comment by TobiHartmann** [comment]
> Memo to myself: there are no trivial performance fixes

I'll copy that memo, it did look harmless at the time.

File: src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp (hunk 1 above)
Comment (TobiHartmann) [reply src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:289]: Thanks for looking at this @adinn. Right, the macro assembler merge magic is nice, I didn't know about it.

**Review by chhagedorn — APPROVED** [review]
Looks good.

**Comment by TobiHartmann** [comment]
Thanks, Christian!

**Pull request closed without merging**